### PR TITLE
Fix health client leak when calling UpdateAllTargets

### PR DESCRIPTION
**Pull request by klauspost**
## Description

When `LoadBucketMetadataHandler` is called and `UpdateAllTargets` gets called. Unless all targets are removed health checks aren't canceled.

Since targets are rebuilt we cancel all.

## Motivation and Context

Avoid memory buildup.

## How to test this PR?

Update remote targets. Observe if clients health goroutines leak.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: cmd/bucket-targets.go [hunk 1]
```diff
@@ -272,17 +272,20 @@ func (sys *BucketTargetSys) UpdateAllTargets(bucket string, tgts *madmin.BucketT
...
-				}
-				delete(sys.arnRemotesMap, t.Arn)
+
+	// Remove existingtarget and arn association
+	if tgts, ok := sys.targetsMap[bucket]; ok {
+		for _, t := range tgts {
+			if tgt, ok := sys.arnRemotesMap[t.Arn]; ok && tgt.healthCancelFn != nil {
+				tgt.healthCancelFn()
 			}
+			delete(sys.arnRemotesMap, t.Arn)
```
Comment (klauspost) [inline cmd/bucket-targets.go:282]: @poornas This was in the original code. Could you verify that Arn's are unique per bucket? I imagine they are, but want to make double sure.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#a2ec5bdc-6a56-4f06-9b7f-7ab55b6f81b3) |
| mint-fs.sh | :x: [more...](#6bfc18cc-4f0f-471d-b6d9-c6d96990f0e8) |
| mint-gateway-s3.sh | :x: [more...](#4762500d-6368-4362-8345-c21379197473) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#3bde91af-32cc-4aae-8948-a5187dd66fe1) |
| mint-pools.sh | :x: [more...](#b9da091b-4c5e-4fee-9e13-4dce016984d3) |
<a name="b9da091b-4c5e-4fee-9e13-4dce016984d3"></a>
### 15167-f4852b7/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s3:31840
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 6 seconds
(5/15) Running awscli tests ... done in 57 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 30 seconds
FAILED in 2 minutes and 15 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 128
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 128
(8/15) Running minio-go tests ... done in 1 minutes and 30 seconds
(9/15) Running minio-java tests ... done in 37 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 6 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 34 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 6 seconds

Executed 14 out of 15 tests successfully.
```
<a name="6bfc18cc-4f0f-471d-b6d9-c6d96990f0e8"></a>
### 15167-f4852b7/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s5:31878
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 52 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 22 seconds
FAILED in 2 minutes and 39 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 141
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 141
(8/15) Running minio-go tests ... done in 1 minutes and 3 seconds
(9/15) Running minio-java tests ... done in 17 seconds
(10/15) Running minio-js tests ... done in 59 seconds
(11/15) Running minio-py tests ... done in 47 seconds
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running versioning tests ... done in 3 minutes and 1 seconds

Executed 14 out of 15 tests successfully.
```
<a name="4762500d-6368-4362-8345-c21379197473"></a>
### 15167-f4852b7/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s3:32189
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 5 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 6 seconds
(5/15) Running awscli tests ... done in 51 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 26 seconds
FAILED in 2 minutes and 57 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 145
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 145
(8/15) Running minio-go tests ... done in 37 seconds
(9/15) Running minio-java tests ... done in 32 seconds
(10/15) Running minio-js tests ... done in 41 seconds
(11/15) Running minio-py tests ... done in 57 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 5 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running versioning tests ... done in 1 seconds

Executed 14 out of 15 tests successfully.
```
<a name="a2ec5bdc-6a56-4f06-9b7f-7ab55b6f81b3"></a>
### 15167-f4852b7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s5:30515
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 7 seconds
(5/15) Running awscli tests ... done in 57 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 37 seconds
FAILED in 2 minutes and 15 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 128
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 128
(8/15) Running minio-go tests ... done in 1 minutes and 16 seconds
(9/15) Running minio-java tests ... done in 33 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 4 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 24 seconds
(12/15) Running s3cmd tests ... done in 26 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 14 out of 15 tests successfully.
```
<a name="3bde91af-32cc-4aae-8948-a5187dd66fe1"></a>
### 15167-f4852b7/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s6:30187
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 7 seconds
(5/15) Running awscli tests ... done in 58 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 34 seconds
FAILED in 1 minutes and 51 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 123
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 123
(8/15) Running minio-go tests ... done in 1 minutes and 11 seconds
(9/15) Running minio-java tests ... done in 31 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 5 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 17 seconds
(12/15) Running s3cmd tests ... done in 23 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#2cbfb258-8fc4-4117-91e5-3b6ab4e7f08a) |
<a name="2cbfb258-8fc4-4117-91e5-3b6ab4e7f08a"></a>
### 15167-f4852b7/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s5:32504
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 5 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 52 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 23 seconds
(8/15) Running minio-dotnet tests ... done in 6 minutes and 12 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 17 seconds
{
  "args": {
    "bucketName": "minio-go-test-fop1rb50jqx5u0o3",
    "expires": 3600000000000,
    "objectName": "minio-go-test-fop1rb50jqx5u0o3unique",
    "reqParams": {
      "response-content-disposition": [
        "attachment; filename=\"test.txt\""
      ]
    }
  },
  "duration": 40875,
  "error": "Get \"http://minio-k8s5:32504/minio-go-test-fop1rb50jqx5u0o3/minio-go-test-fop1rb50jqx5u0o3unique?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20220624%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220624T164106Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3D%22test.txt%22&X-Amz-Signature=024d12766d242c41c98db4c835764defba5ca93147d509610e63d2e7968fb1de\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)",
  "function": "PresignedGetObject(bucketName, objectName, expires, reqParams)",
  "message": "PresignedGetObject response incorrect",
  "name": "minio-go: testFunctional",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 20 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 0 seconds
(11/15) Running minio-py tests ... done in 48 seconds
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 1 seconds
(14/15) Running versioning tests ... done in 3 minutes and 2 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-targets.go [hunk 2]
```diff
@@ -272,17 +272,20 @@ func (sys *BucketTargetSys) UpdateAllTargets(bucket string, tgts *madmin.BucketT
...
 		}
 		delete(sys.targetsMap, bucket)
+	}
+
+	// No need for more if not adding anything
+	if tgts == nil || tgts.Empty() {
```
Comment (krisis) [inline cmd/bucket-targets.go:288]: [Optional] With the simplification brought about in this PR we can remove the following [block](https://github.com/minio/minio/pull/15167/files#diff-4cb845cd18a0f200b38b18c6ae381eb5a9572e944482e62b18430b6ee968fe3fR293).
```go
if len(tgts.Targets) > 0 {
    sys.targetsMap[bucket] = tgts.Targets
}
```

**Review by krisis — APPROVED** [review]

**Review by poornas — APPROVED** [review]

File: cmd/bucket-targets.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-targets.go:282]: they will be unique for the bucket

File: cmd/bucket-targets.go (hunk 2 above)
Comment (klauspost) [reply cmd/bucket-targets.go:288]: I agree. Didn't want to accidentally break something, so kept it to a minimum.

**Pull request merged**